Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    // Copy the index since _Py_Identifier.index is declared as volatile
+    Py_ssize_t index = id->index;
+    if (index < 0) {
+        struct _Py_unicode_runtime_ids *rt_ids = &interp->runtime->unicode_ids;
+
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
+        // Check again to detect concurrent access. Another thread can have
+        // initialized the index while this thread waited for the lock.
+        //
+        // _Py_Identifier.index is declared with volatile keyword to ensure
+        // that the index is read again here (prevent compiler optimization).
+        // Use volatile instead of an atomic variable since pycore_atomic.h is
+        // an internal header file, and stdatomic.h is not yet widely adopted.
+        if (id->index < 0) {
+            id->index = rt_ids->next_index;
+            rt_ids->next_index++;
+        }
+        PyThread_release_lock(rt_ids->lock);
+        index = id->index;
+    }
+    assert(index >= 0);
+
+    PyObject *object;
+    if (index < ids->size) {
+        object = ids->array[index];
```

Comment:
Humm, and the array should be volatile and atomic too if we read it without locking.

Reply:
Again, it's guarded by the GIL.